### PR TITLE
test: Add epochs to MBT

### PR DESCRIPTION
**Pull request by p-offtermatt**
## Description

Closes: #1669 

Adds epochs to MBT.

Because the MBT driver produces a ton of extra blocks (due to the use of ibctesting, which mainly produces extra blocks when updating clients) I get around ~15 extra blocks per block in the model (but this is not a constant, depends on how many consumers are running, there are more extra blocks for starting consumers, etc)

So I solve epochs in a slightly hacky way:
I set the blocksPerEpoch parameter to a very large value, which will not fill up during normal execution.
Then, when an epoch in the model ends, I detect this and produce many blocks to actually end the epoch in the system.
There is a "safety breaker", i.e. I check that the system does not enter a new epoch unless the model does, too, and give a warning to increase blocksPerEpoch if this is violated.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [x] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

File: tests/mbt/driver/mbt_test.go [hunk 1]
```diff
@@ -523,6 +558,9 @@ func ComparePacketQueues(
 	consumer string,
 	timeOffset time.Time,
 ) {
+	if consumer == "consumer1" {
```
Comment (insumity) [inline tests/mbt/driver/mbt_test.go:561]: Why is this needed here?

File: tests/mbt/driver/mbt_test.go [hunk 2]
```diff
@@ -200,8 +208,16 @@ func RunItfTrace(t *testing.T, path string) {
 	t.Log("Reading the trace...")
 
 	for index, state := range trace.States {
+		t.Log("Height modulo epoch length:", driver.providerChain().CurrentHeader.Height%blocksPerEpoch)
+		t.Log("Model height:", ProviderHeight(state.VarValues["currentState"].Value.(itf.MapExprType))%modelBlocksPerEpoch)
```
Comment (insumity) [inline tests/mbt/driver/mbt_test.go:212]: nit: similar to the line above maybe say `"Model height modulo model epoch length"`

File: tests/mbt/driver/mbt_test.go [hunk 3]
```diff
@@ -200,8 +208,16 @@ func RunItfTrace(t *testing.T, path string) {
...
 		t.Logf("Reading state %v of trace %v", index, path)
 
+		// store the height of the provider state before each step.
+		// The height should only pass an epoch when there passes an epoch in the model, too,
```
Comment (insumity) [inline tests/mbt/driver/mbt_test.go:216]:
```suggestion
		// The height should only pass an epoch when it passes an epoch in the model, too,
```

File: tests/mbt/driver/mbt_test.go [hunk 4]
```diff
@@ -239,22 +255,33 @@ func RunItfTrace(t *testing.T, path string) {
...
+
+				// produce blocks until the next block ends the epoch
+				for i := heightInEpoch; i < blocksPerEpoch; i++ {
+					driver.endAndBeginBlock("provider", 1*time.Nanosecond)
+				}
+			}
+
 			// we need at least 2 blocks, because for a packet sent at height H, the receiving chain
 			// needs a header of height H+1 to accept the packet
 			// so, we do `blocksPerEpoch` time advancements with a very small increment,
```
Comment (insumity) [inline tests/mbt/driver/mbt_test.go:276]: Remove the `// so, we do ... ` comment.

File: tests/mbt/driver/setup.go [hunk 5]
```diff
@@ -397,21 +397,15 @@ func (s *Driver) ConfigureNewPath(consumerChain, providerChain *ibctesting.TestC
 	// their channel, and are ready for anything to happen.
 	s.consumerKeeper(consumerChainId).SetProviderChannel(s.ctx(consumerChainId), consumerEndPoint.ChannelID)
 
-	// Commit a block on both chains, giving us two committed headers from
-	// the same time and height. This is the starting point for all our
-	// data driven testing.
+	// Commit a block on the consumer to get a header
```
Comment (insumity) [inline tests/mbt/driver/setup.go:400]: Why do we have this change here from `provider` to `consumer`?

File: tests/mbt/model/ccv.qnt [hunk 6]
```diff
@@ -86,11 +89,8 @@ module ccv_types {
             // Stores VscPackets which have been sent but where the provider has *not received a response yet*.
             sentVscPacketsToConsumer: Chain -> List[VscPacket],
 
-            // stores whether, in this block, the validator set has changed.
-            // this is needed because the validator set might be considered to have changed, even though
-            // it is still technically identical at our level of abstraction, e.g. a validator power change on the provider
-            // might leave the validator set the same because a delegation and undelegation cancel each other out.
-            providerValidatorSetChangedInThisBlock: bool,
+            // stores for which consumer chains, in this epoch, the validator set is considered to have changed and we thus need to send a vscPacket to the consumer chains.
```
Comment (insumity) [inline tests/mbt/model/ccv.qnt:92]: nit:
```suggestion
            // stores for which consumer chains, in this epoch, the validator set is considered to have changed and we thus need to send a VscPacket to the consumer chains.
```

File: tests/mbt/model/ccv_model.qnt [hunk 7]
```diff
@@ -517,9 +541,10 @@ module ccv_model {
                     assert(currentState.providerState.chainState.votingPowerHistory == List(InitialValidatorSet.put("node1", 150), InitialValidatorSet)),
                     // change voting power on provider again
                     VotingPowerChange("node1", 50).then(
-                        // end another block
-                        EndAndBeginBlockForProvider(1 * Second, Set(), Set())
-                    ).then(
+                        // end the epoch
+                        EndProviderEpoch(3 * Second, Set(), Set())
```
Comment (insumity) [inline tests/mbt/model/ccv_model.qnt:545]: Why do we have 3 seconds here instead of 1 as is done in `EndProviderEpoch` below?

File: tests/mbt/driver/mbt_test.go (hunk 1 above)
Comment (p-offtermatt) [reply tests/mbt/driver/mbt_test.go:561]: Oops, a legacy left over from debugging. Thanks for catching!

File: tests/mbt/driver/setup.go (hunk 5 above)
Comment (p-offtermatt) [reply tests/mbt/driver/setup.go:400]: I removed an extra block production for the provider here because I thought I could make it so model and system heights would line up, if I managed to remove this (but it turns out there are waaay more places where blocks are produced than I thought), so the comment needed to change, too

File: tests/mbt/model/ccv_model.qnt (hunk 7 above)
Comment (p-offtermatt) [reply tests/mbt/model/ccv_model.qnt:545]: "Historical reasons", i.e. a relic from trying things while debugging :)
good catch, adjusted this!

File: tests/mbt/driver/setup.go (hunk 5 above)
Comment (p-offtermatt) [reply tests/mbt/driver/setup.go:400]: Turns out I was able to remove the extra block for the consumer, too, so now the comment is completely gone :)

**Review by insumity — APPROVED** [review]

File: tests/mbt/model/ccv_model.qnt [hunk 8]
```diff
@@ -319,16 +340,19 @@ module ccv_model {
...
+
+    val CurrentBlockEndsEpoch = currentState.providerState.chainState.currentBlockHeight % epochLength == 0
+
     // Any update in the power of a validator on the provider
     // MUST be present in a ValidatorSetChangePacket that is sent to all registered consumer chains
     val ValUpdatePrecondition = trace[trace.length()-1].kind == "EndAndBeginBlockForProvider"
     val ValidatorUpdatesArePropagatedInv = 
-        // when the provider has just entered a validator set into a block...
-        ValUpdatePrecondition and currentState.providerState.providerValidatorSetChangedInThisBlock
+        // when the an epoch ends and the provider has just entered a validator set into a block...
```
Comment (sainoe) [inline tests/mbt/model/ccv_model.qnt:350]:
```suggestion
        // when an epoch ends and the provider has just entered a validator set into a block...
```

File: tests/mbt/model/ccv_model.qnt [hunk 9]
```diff
@@ -235,6 +239,23 @@ module ccv_model {
...
+    // The rest of the blocks will not start or stop any consumers, and will advance time by 1 second each.
+    run EndProviderEpoch(
+        timeAdvancement: Time,
+        consumersToStart: Set[Chain],
+        consumersToStop: Set[Chain]
+    ): bool =
+        epochLength.reps(
+            i => 
+            if (i == 0)
+            EndAndBeginBlockForProvider(timeAdvancement-((epochLength-1)*Second), consumersToStart, consumersToStop)
```
Comment (mpoke) [inline tests/mbt/model/ccv_model.qnt:254]: What if `timeAdvancement-((epochLength-1)*Second)` is negative?

File: tests/mbt/model/ccv_model.qnt [hunk 10]
```diff
@@ -235,6 +239,23 @@ module ccv_model {
         stepCommon
     }
 
+    // Runs epochLength many blocks on the provider.
+    // The first block will start all consumers in consumersToStart and stop all consumers in consumersToStop,
+    // and advance time by timeAdvancement - ((epochLength-1) * Seconds).
+    // The rest of the blocks will not start or stop any consumers, and will advance time by 1 second each.
+    run EndProviderEpoch(
```
Comment (mpoke) [inline tests/mbt/model/ccv_model.qnt:246]: This is only used for tests, right? Could you rename it accordingly or make it clear in the comment.

**Review by mpoke — APPROVED** [review]
LGTM. Great work @p-offtermatt

**Review by sainoe — APPROVED** [review]
LGTM!
